File: publication/PrimeSynthesis.kt
import java.math.BigInteger
import kotlin.collections.indices
import kotlin.sequences.toList

/**
 * Generates all unique binary partitions of a set.
 */
private fun <T> binaryPartitions(list: List<T>): Sequence<Pair<List<T>, List<T>>> =
    kotlin.sequences.sequence {
        if (list.size < 2) return@sequence
        val first = list[0]
        val rest = list.subList(1, list.size)
        val numPartitions = 1L shl rest.size

        for (i in 1 until numPartitions) {
            val left = mutableListOf<T>()
            val right = mutableListOf<T>(first)
            for (j in rest.indices) {
                if ((i shr j) and 1L == 1L) left.add(rest[j]) else right.add(rest[j])
            }
            yield(left to right)
        }
    }

/**
 * Generate all exponent combinations (odometer approach)
 */
private fun exponentCombinations(n: Int, maxExp: Int): Sequence<List<Int>> =
    kotlin.sequences.sequence {
        if (n == 0) return@sequence
        val indices = IntArray(n) { 1 }
        while (true) {
            yield(indices.toList())
            var i = n - 1
            while (i >= 0 && indices[i] == maxExp) {
                indices[i] = 1
                i--
            }
            if (i < 0) break
            indices[i]++
        }
    }

/**
 * Synthesizes primes using arbitrary-precision arithmetic to prevent overflow.
 */
private fun computePrimes(seeds: List<BigInteger>, maxExp: Int): Sequence<BigInteger> {
    if (seeds.isEmpty()) return kotlin.sequences.emptySequence()
    val maxP = seeds.maxOrNull() ?: return kotlin.sequences.emptySequence()
    val minR = maxP.add(BigInteger.ONE)
    val maxR = maxP.multiply(maxP)

    return sequence {
        for ((left, right) in binaryPartitions(seeds)) {
            val totalSize = left.size + right.size
            for (expVector in exponentCombinations(totalSize, maxExp)) {
                val leftExps = expVector.take(left.size)
                val rightExps = expVector.takeLast(right.size)

                val leftProd = left.zip(leftExps).map { (p, e) -> p.pow(e) }
                    .reduce { acc, b -> acc.multiply(b) }
                val rightProd = right.zip(rightExps).map { (p, e) -> p.pow(e) }
                    .reduce { acc, b -> acc.multiply(b) }

                val sumSigma = leftProd.add(rightProd)
                if (sumSigma >= minR && sumSigma <= maxR) yield(sumSigma)

                val diffSigma = leftProd.subtract(rightProd).abs()
                if (diffSigma >= minR && diffSigma <= maxR) yield(diffSigma)
            }
        }
    }
}

fun main() {
    var currentMaxExp = 1
    val maxIterations = 5
    var current = listOf(BigInteger.ONE, BigInteger.valueOf(2))
    var accPrimes = emptyList<BigInteger>()

    for (i in 1..maxIterations) {
        val found = computePrimes(current, currentMaxExp++).toList()
        val distinctFound = found.sorted().distinct()

        val currentSet = current.toSet()
        val diff = distinctFound.filter { it !in currentSet }
        val next = if (diff.isEmpty()) current else current + diff.min()

        accPrimes = accPrimes + distinctFound
        current = next
    }
    println("Discovered primes: ${(listOf(BigInteger.valueOf(2)) + accPrimes).distinct().sorted()}")
}

File: publication/prime_synthesis.py
from itertools import product as itertools_product


def binary_partitions(lst):
    """Generates all unique binary partitions of a list."""
    if len(lst) < 2:
        return
    first = lst[0]
    rest = lst[1:]
    num_partitions = 1 << len(rest)
    for i in range(1, num_partitions):
        left = []
        right = [first]
        for j in range(len(rest)):
            if (i >> j) & 1:
                left.append(rest[j])
            else:
                right.append(rest[j])
        yield (left, right)


def exponent_combinations(n, max_exp):
    """Generate all exponent vectors (odometer approach)."""
    if n == 0:
        return
    indices = [1] * n
    while True:
        yield list(indices)
        i = n - 1
        while i >= 0 and indices[i] == max_exp:
            indices[i] = 1
            i -= 1
        if i < 0:
            break
        indices[i] += 1


def compute_primes(seeds, max_exp):
    """Synthesizes candidates within the quadratic window (max_p, max_p^2]."""
    if not seeds:
        return []
    max_p = max(seeds)
    min_r = max_p + 1
    max_r = max_p * max_p
    results = []
    for left, right in binary_partitions(seeds):
        total_size = len(left) + len(right)
        for exp_vector in exponent_combinations(total_size, max_exp):
            left_exps = exp_vector[:len(left)]
            right_exps = exp_vector[len(left):]
            left_prod = 1
            for p, e in zip(left, left_exps):
                left_prod *= p ** e
            right_prod = 1
            for p, e in zip(right, right_exps):
                right_prod *= p ** e
            sum_sigma = left_prod + right_prod
            if min_r <= sum_sigma <= max_r:
                results.append(sum_sigma)
            diff_sigma = abs(left_prod - right_prod)
            if min_r <= diff_sigma <= max_r:
                results.append(diff_sigma)
    return results


def main():
    current_max_exp = 1
    max_iterations = 5
    current = [1, 2]
    acc_primes = []

    for i in range(1, max_iterations + 1):
        found = compute_primes(current, current_max_exp)
        current_max_exp += 1
        distinct_found = sorted(set(found))
        current_set = set(current)
        diff = [x for x in distinct_found if x not in current_set]
        if diff:
            current = current + [min(diff)]
        acc_primes.extend(distinct_found)

    all_primes = sorted(set([2] + acc_primes))
    print(f"Discovered primes: {all_primes}")


if __name__ == "__main__":
    main()